[DataViews]: Support dual rendering mode (controlled props or free composition)

retrofox opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Closes WOOA7S-124

## Proposed Changes

* Refactor `<DataViews />` to support both `controlled` usage (as before) and a new `free-composition` mode, where consumers can pass `children` and compose their own layout.
* Introduce a new `defaultUI` block internally, which is conditionally rendered if no `children` are passed.
* Add a new Storybook example to illustrate the free composition approach.

## Why are these changes being made?

This update enables a more flexible usage of `<DataViews />`, allowing consumers to use the default internal UI or compose their own layout and behavior using children.

This pattern is helpful for cases where we want to build completely custom layouts while still relying on the shared context logic for selections, filtering, pagination, etc.

```ts
<DataViews data={ myData }>
  <DataViews.Search />
  <DataViews.Filters />
  <DataViews.ViewConfig />
  <MyCustomComponent />
  <DataViews.Layout />
</DataViews>
```

## Testing Instructions

1) Start the Storybook app

```
yarn dataviews:storybook:start
```

2) Run the existing stories and verify that the controlled version of `<DataViews />` still behaves as expected.

3) Open the new `FreeComposition` story:
   - Confirm that it renders using the context-based data.
   - Confirm no runtime or render errors.
   - Check that you can access filtered/paginated data from context or via the `filterSortAndPaginate()` utility.

<img width="920" alt="image" src="https://github.com/user-attachments/assets/8690d8c3-c8f5-49f1-ab1a-c8668ef72148" />


## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

retrofox committed:
Refactor DataViews to support controlled and free composition
Improves type safety and handles optional props

retrofox committed:
export DataViewsContextType type

retrofox committed:
add Free composition story

retrofox committed:
Improve PlanetOverview component props
Pass data directly instead of using context

retrofox commented:
TypeScript error occurs [here](https://github.com/Automattic/wp-calypso/blob/8db910887432f73b33b6b583cd2f21b5d1c9b551/packages/dataviews/src/components/dataviews/stories/index.story.tsx#L66):

> Parameter 'item' implicitly has an 'any' type.ts(7006)

```ts
onClickItem={ ( item ) => {
```

This happens because DataViewsProps is now a discriminated union type, supporting both controlled usage and free composition.
However, since there’s no explicit discriminant (type or similar), TypeScript cannot reliably narrow prop types in consumer code when inferring function parameter types, like here, in `onClickItem`.

As a result, `item` is implicitly typed as `any`. This can be resolved locally by explicitly annotating the parameter (e.g., item: `SpaceObject`), but it’s not ideal from a DX perspective.


 With this new API design, TypeScript may be unable to infer the correct type for some function parameters in consumer code, especially for callbacks like onClickItem.
Developers might see an error like Parameter 'item' implicitly has an 'any' type.ts(7006) and need to annotate the parameter manually. This is a known trade-off of supporting both controlled and free composition through a union type.

retrofox committed:
Rename DataViewsWithChildren to DataViewsCompositionProps

retrofox committed:
fix consumer TS issue

retrofox committed:
fix FreeComposition story
Pick the planets from the fixtures data

retrofox commented on packages/dataviews/src/components/dataviews/stories/index.story.tsx:
```diff
@@ -105,3 +122,63 @@ export const FieldsNoSortableNoHidable = () => {
+
+	return (
+		<DataViews>
+			<PlanetOverview planets={ planets } />
```
Here is where we'll add the DataViews subcomponents:

```ts
        <DataViews.Search />
        <DataViews.Filters />
        <DataViews.BulkActions />
        <DataViews.LayoutSwitcher />
        <DataViews.ViewConfig />
```

oandregal replied:
Any children should access data via the context, not via props. This is important because data can be filtered, sorted, or paginated. Each component shouldn't need to reimplement this logic by itself. See how this [alternative approach](https://github.com/Automattic/wp-calypso/pull/102996#issuecomment-2841266284) handles data.

retrofox replied:
That's partially correct from my PoV. We can rephrase it like:

~you're right :-)~

retrofox replied:
The processed data (sorted, filtered, etc) is accessible through the public `filterSortAndPaginate()` helper

retrofox commented on packages/dataviews/src/components/dataviews/stories/index.story.tsx:
```diff
@@ -46,7 +63,7 @@ export const Default = () => {
 			onChangeView={ setView }
 			actions={ actions }
-			onClickItem={ ( item ) => {
+			onClickItem={ ( item: SpaceObject ) => {
```
Here is where the TS error occurs:

> Parameter 'item' implicitly has an 'any' type.ts(7006)

This happens because DataViewsProps is now a [discriminated union type](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions), supporting both controlled usage and free composition.
However, since there’s no explicit discriminant (type or similar), TypeScript cannot reliably narrow prop types in consumer code when inferring function parameter types, like here, in `onClickItem`.

As a result, `item` is implicitly typed as `any`. This can be resolved locally by explicitly annotating the parameter (e.g., item: `SpaceObject`), but it’s not ideal from a DX perspective.

With this new API design, TypeScript may be unable to infer the correct type for some function parameters in consumer code, especially for callbacks like `onClickItem`.

Developers might see an error like `Parameter 'item' implicitly has an 'any' type.ts(7006)` and need to annotate the parameter manually. This is a known trade-off of supporting both controlled and free composition through a union type.

cc @youknowriad and @jsnajdr in case you have any tips to address this?

retrofox replied:
Fixed after setting a few component props as optional.

retrofox committed:
improve free composition story jsdoc

louwie17 commented on packages/dataviews/src/components/dataviews/index.tsx:
```diff
@@ -53,29 +58,42 @@ type DataViewsProps< Item > = {
+export type DataViewsCompositionProps< Item > = {
+	children: ReactNode;
+	data?: Item[];
+} & Partial< Omit< DataViewsControlledProps< Item >, 'children' | 'data' > >;
```
Any specific reason for not just adding `children?: ReactNode` to `DataViewsProps` and keeping the rest the same?
I understand that maybe not all these items are necessary anymore as a consumer may not use all aspects of the nested DataViews structure, but if they do, they still require all the above props to be passed in.

IMO, making everything else optional when `children` is passed in convolutes the types. I would suggest just adding `children?: ReactNode`.

youknowriad replied:
Same here. I think everything else should not be optional actually. There's just the "header" prop that don't make much sense in the context of the "provider". All the rest seems like it should remain as is.

retrofox replied:
I’m sorry, but this isn’t how TypeScript is supposed to work. Its primary purpose is to help us anticipate problems during static analysis, *before* we even run the code.

If we make all props optional, what’s the point of having types? We’ll be able to do something like:

```typescript
<DataViews>
   <p>Everything is okay!</p>
</DataViews>
```

TypeScript will be happy, but then at runtime…

```
normalize-fields.ts:33 Uncaught TypeError: Cannot read properties of undefined (reading 'map')
    at normalizeFields (normalize-fields.ts:33:1)
    at index.tsx:104:1
    at DataViews (index.tsx:104:1)
```

Clearly, we’re not using TypeScript as intended. That leads me to the question: if we allow this, why use TypeScript at all instead of plain JavaScript?

youknowriad replied:
I think what we're saying is not to make everything optional, but to just make `children` optional because that's our intention here. 

I think what @louwie17 is saying is that you made everything optional by using `Partial`

retrofox replied:
Ah, got it, pardon!. thanks for clarifying!

But in that case, we’re just hitting the other side of the same coin: if we don’t make the props optional, then TypeScript will require all of them, even in the free composition mode, where many of them might not make sense.

For the same example above, it will throw tons of TypeScript errors, even though this is a valid and intentional use case.

That’s actually the reason we introduced the two separate types, to reflect the two modes DataViews supports: the controlled mode (the current one), and the free composition mode (using children).

The idea is to align the typing with how the component is being used, based on whether children is present or not.

youknowriad replied:
>The idea is to align the typing with how the component is being used, based on whether children is present or not. 

Yes, it seems you consider the props of the provider different than the props of the regular DataViews. This seems to be the main difference here. Why do you think so, why should these props be optional for the provider but mandatory for DataViews?

jsnajdr replied:
What if we created a new React component for this free composition mode? `<DataViews>` is controlled, `<DataViewsContainer>` is a container for children that specify the free-form layout.

I think it's better that squeezing everything into one component, with two wildly different sets of props.

Would that simplify the types and avoid the problems with `onClickItem`? If the complexity of the types gets out of control, that's a hint that maybe we should refactor the design.

retrofox replied:
> Why do you think so, why should these props be optional for the provider but mandatory for DataViews?

Because free composition changes the rules around which props are required.

For example, `children` and `data` are clearly required, and there are likely other props that should also be required in this mode, but that’s something we still need to analyze carefully.

At the same time, depending on the composition, some props might not be needed. For instance, if we’re not rendering a `<Pagination />` component, do we really need to define the `paginationInfo` prop?

retrofox replied:
@jsnajdr I think this Automattic/wp-calypso#102987, which is aligned with your suggestion. I took that approach precisely to address the `onClickItem` issue.

The name is `<DataViewsProvider />`,

youknowriad replied:
> At the same time, depending on the composition, some props might not be needed. For instance, if we’re not rendering a <Pagination /> component, do we really need to define the paginationInfo prop?

For me paginationInfo has more to do about the "data" and not about whether pagination is rendered or not. In other words, I don't see this as something to differentiate the props between controlled UI and uncontrolled UI. I can be onboard with making this prop optional but I would do so for all use-cases not just the "uncontrolled UI".

retrofox replied:
Okay, sounds good to me. Let’s iterate on it then

youknowriad commented on packages/dataviews/src/components/dataviews/index.tsx:
```diff
@@ -100,39 +120,59 @@ export default function DataViews< Item >( {
+		containerWidth,
+	};
+
+	if ( children ) {
```
We can avoid the condition entirely by just putting the fallback DataViews UI in a `DefaultUI ` component and doing `children = <DefaultUI />` or something.

oandregal commented:
Trying to find the minimal change that works, I found a oneliner. Add this line to `packages/dataviews/src/index.ts`:

```js
export { default as DataViewsContext } from './components/dataviews-context';
```

Then, we can do the following:

```js
export const FreeComposition = () => {
	const view = DEFAULT_VIEW;
	const { data: shownData, paginationInfo } = useMemo( () => {
		return filterSortAndPaginate( data, view, fields );
	}, [ view ] );

	const getItemId = ( item: { id: string } ) => item.id;
	const [ openedFilter, setOpenedFilter ] = useState< string | null >( null );
	const isItemClickable = () => true;
	const containerWidth = 0; // TODO: this should be updated upon a resize event on a wrapper defined by the consumer.

	return (
		<DataViewsContext.Provider
			value={ {
				getItemId,
				data: shownData,
				paginationInfo,
				view,
				onChangeView: () => {},
				fields: [],
				selection: [],
				onChangeSelection: () => {},
				openedFilter,
				setOpenedFilter,
				isItemClickable,
				containerWidth,
			} }
		>
			<PlanetOverview />
		</DataViewsContext.Provider>
	);
};
```

Note that `PlanetOverview` should access data via the context (`const { data } = useContext( DataViewsContext );`) and not via props.

Doing this exercise also highlights a couple of things that are hidden with the "dual function" of DataViews:

- The context could absorb more defaults that are now part of `DataViews` component (selection, defaults such as getItemId, etc.).
- The story needs to use filters to demonstrate how that'd work. For example, there are two missing pieces of data that are important for consumers, but we don't pass via context: `isShowingFilter`, `setIsShowingFilter`.
- The `containerWidth` needs a wrapper component to calculate the width from ([see](https://github.com/Automattic/wp-calypso/blob/trunk/packages/dataviews/src/components/dataviews/index.tsx#L80)).

I think I'd be more comfortable starting with this simple approach and then iterate in subsequent PRs to address those things bit by bit until we have a context that absorbs all (most?).

retrofox commented:
> Note that PlanetOverview should access data via the context (const { data } = useContext( DataViewsContext );) and not via props.

In that case, I think we should expose the context, right? The story reflects a real case, and I assumed we didn't want to make it public. That's why I switched the approach, passing the props down to the children component.

More context about whether making the context public is [a good idea here](https://github.com/Automattic/wp-calypso/pull/102967#issuecomment-2839836759).

> Doing this exercise also highlights a couple of things that are hidden with the "dual function" of DataViews:

> The context could absorb more defaults that are now part of DataViews component (selection, defaults such as getItemId, etc.).

It's a nice one.

> The story needs to use filters to demonstrate how that'd work. For example, there are two missing pieces of data that are important for consumers, ...

Agree. I added a jsdoc block about it [here](https://github.com/Automattic/wp-calypso/blob/33b951a161fe33c2d4e802a08b4396dcd1b49b42/packages/dataviews/src/components/dataviews/stories/index.story.tsx#L173-L187).

> but we don't pass via context: isShowingFilter, setIsShowingFilter.

Do you think we should define these props as required?

> The containerWidth needs a wrapper component to calculate the width from ([see](https://github.com/Automattic/wp-calypso/blob/trunk/packages/dataviews/src/components/dataviews/index.tsx#L80)).

👍 

> I think I'd be more comfortable starting with this simple approach and then iterate in subsequent PRs to address those things bit by bit until we have a context that absorbs all (most?).

💯

retrofox commented:
> Trying to find the minimal change that works, I found a oneliner. Add this line to packages/dataviews/src/index.ts:
> ```
> export { default as DataViewsContext } from './components/dataviews-context';
> ```

If we expose the context, I suggest doing it via a custom hook. Going to clean https://github.com/Automattic/wp-calypso/pull/102967 and in the wors of cases we close it :-)

oandregal commented:
> If we expose the context, I suggest doing it via a custom hook. 

We should have a mindset of doing the minimal change that works, for the reasons shared in the other PRs (sync upstream), so I'd rather go with `useContext( DataViewsContext )`.

retrofox commented:
<!-- linear:replyTo:2841035751 -->

> Then, we can do the following:

Are you suggesting exposing a different component, because that's what the Automattic/wp-calypso#102987 does, with a different name 😀 

[context](https://github.com/Automattic/wp-calypso/pull/102987#pullrequestreview-2805036710) why I created this new approach.

retrofox commented:
> Trying to find the minimal change that works, I found a oneliner. Add this line to `packages/dataviews/src/index.ts`:

The changes introduced by this [other PR are mostly additions](https://github.com/Automattic/wp-calypso/pull/102987/files), avoiding as much as possible modifications in the base code.

retrofox commented:
<!-- linear:replyTo:2841035751 -->

> Personally, I think this PR is the right approach. It's true that it creates a bigger diff than a separate component but I think it's the right way to do it if we had to implement it in the upstream package.

I agree. We should be ready to accept significant changes compared to the core version.

> We should have a mindset of doing the minimal change that works, for the reasons shared in the other PRs (sync upstream), so I'd rather go with `useContext( DataViewsContext )`.

Yes, I’m aware of that, but I still think it might be worth it.

[I’ve updated the PR](https://github.com/Automattic/wp-calypso/pull/102967) and its description. It only changes one line compared to core.

Please take a look when you get the chance 🙇‍♂️

retrofox committed:
data is not optional

youknowriad commented on packages/dataviews/src/components/dataviews/index.tsx:
```diff
@@ -27,9 +27,14 @@ import { normalizeFields } from '../../normalize-fields';
 type ItemWithId = { id: string };
 
-type DataViewsProps< Item > = {
+export type DataViewsControlledProps< Item > = {
```
Why are we exporting this externally?

retrofox replied:
It isn't exposed by package but to be used internally. We use it to deal with the union type. Something that we can revert when removing the two prop types for controlled and free-composition behavior.

louwie17 commented on packages/dataviews/src/components/dataviews/index.tsx:
```diff
@@ -53,29 +58,42 @@ type DataViewsProps< Item > = {
 	? { getItemId?: ( item: Item ) => string }
 	: { getItemId: ( item: Item ) => string } );
 
+export type DataViewsCompositionProps< Item > = {
```
Instead using `Omit` and `Partial`, it may be easier to create a `DataViewsBaseProps` type that includes the mandatory items.
Than create the composition and controlled type like so ( example ):
```
type DataViewsCompositionProps< Item > = {
  children: ReactNode;
} & DataViewsBaseProps< Item >;

type DataViewsControlledProps< Item > = {
  header?: ReactNode;
  // Whatever else is needed for the controlled version.
} & DataViewsBaseProps< Item >;
```

retrofox replied:
Not aligned with the [minimal change](https://github.com/Automattic/wp-calypso/pull/102996#issuecomment-2841332366). We'll change this.

louwie17 replied:
Ok, that's also fine, in that case I do think we can update the below as the `data` type is the same now and `children` doesn't exist in `DataViewsControlledProps` :
```
type DataViewsCompositionProps< Item > = {
  children: ReactNode;
} & Omit< DataViewsControlledProps< Item >, 'header' >
```

louwie17 replied:
> Not aligned with the https://github.com/Automattic/wp-calypso/pull/102996#issuecomment-2841332366. We'll change this.

Having said the above, given we are already renaming this to `DataViewsCompositionProps`, changing it to `DataViewsBaseProps` and removing a couple props, still feels pretty minimal to me :)
But I am ok with sticking with the above.

retrofox committed:
Simplify DataViews props type definition
Unifies controlled and composition modes under single type

retrofox committed:
add getItemId prop in Free Composition story

retrofox committed:
Add data filtering to FreeComposition story
Demonstrates proper data handling in headless mode

retrofox committed:
Remove explicit context type annotation
Relies on type inference for better maintainability

retrofox committed:
simplify

retrofox committed:
fix optional props issues in Context

retrofox committed:
for the sake of the minor changes

retrofox committed:
remove explicity cast in onClickItem

retrofox committed:
Extract default UI to variable and simplify rendering
Improves code organization and reduces duplication

retrofox committed:
changelog

retrofox commented:
I’ve updated the PR. Key changes:

* Removed the separate types for controlled and free-composition modes
* Added default handling for the component prop
* That fixed the lingering TS issues
* Introduced a DefaultUI constant to simplify rendering
* The FreeComposition story now works as expected
* PR description updated

I think we’re in good shape to move forward. I chose clean code over minimal changes.

cc @youknowriad @oandregal

retrofox commented:
There are still quite a few TS issues across the components. From what I’ve seen, I don’t think we can solve them all from a single place like the Context or the core DataViews. They’ll likely need to be addressed gradually and locally in each component.

I’d like to confirm how we want to approach this moving forward.

retrofox commented:
the [Code style failing tests](https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Check_Code_Style_Branch/14816819?expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true&showLog=14816819_918_413&logView=flowAware) shows three are many issues to fix. 

I have to say the [DataViews]: introduce DataViewsContainer component #102987 approach seems to be more convenient, because of:

1. It doesn't introduce significant changes, aligned with the idea of _minor changes_ against core. Just a few lines
2. It does not introduce backward compatibility issues

The price to pay is that we expose a new component.

youknowriad commented on packages/dataviews/src/components/dataviews-context/index.ts:
```diff
@@ -11,18 +11,18 @@ import type { SetSelection } from '../../private-types';
-	view: View;
-	onChangeView: ( view: View ) => void;
+	view?: View;
+	onChangeView?: ( view: View ) => void;
```
For me view should be mandatory, onChangeView could potentially be optional but that requires deep changes to the DataViews layouts to disable sorting, fitlering... so for now, I think we should keep all the existing props untouched. (including paginationInfo, onChangeSelection)

Not rendering the DataViews layout (table, or list... ) doesn't mean these props should be made optional, it just means that the layout will be rendered manually at the right position. The layouts still expect these things to be defined.

retrofox replied:
can we make it required by the Context, but optional for the component passing a default `view = { type: LAYOUT_TABLE }` prop? https://github.com/Automattic/wp-calypso/pull/102996/commits/b785dd4320ff0fb39d44acdcb111f8d72467b0c7

It ensures a defined `view` in the context, but is optional for the consumer when creating DataViews instances

retrofox replied:
It's similar to the `fields[]` prop, [required by the context](https://github.com/Automattic/wp-calypso/blob/c5ec43891c6abec612d9fa1504a9dadb9d1a0b10/packages/dataviews/src/components/dataviews-context/index.ts#L16), but optional by the component with a default empty array value.

youknowriad replied:
For me a `DataViews` component without fields or without view doesn't really make sense. So I'd return the question to you, why do you want a DataViews component without view or fields, what would you use that for?

retrofox committed:
Make view prop required in DataViews component
Ensures consistent state management

retrofox commented on packages/dataviews/src/components/dataviews/index.tsx:
```diff
@@ -25,30 +25,31 @@ import DataViewsSearch from '../dataviews-search';
-import type { SelectionOrUpdater } from '../../private-types';
+import type { SelectionOrUpdater, SetSelection } from '../../private-types';
 
 type ItemWithId = { id: string };
```
@youknowriad could we accept item `id` being a Number?

```suggestion
type ItemWithId = { id: string | number };
```

youknowriad replied:
I'll defer to @oandregal and others. But do you want the "id" to be a number? If your ids are number in your objects, you can always provide `getItemId` prop and return `String( item.id )`

youknowriad replied:
Also how does this relate to the current PR?

retrofox replied:
Just wondering because when the IDs are `string[]`, the getItemId gets optional. The fixture we use for the stories, the IDs are `numbers []`, making the `getItemId()` not optional.

retrofox committed:
Add required view prop to FreeComposition story
Improves type safety and adds documentation

retrofox committed:
set de defautlt view for the component

retrofox committed:
rollback defining optional props

retrofox committed:
clean

youknowriad commented on packages/dataviews/src/components/index.ts:
```diff
@@ -0,0 +1,16 @@
+const DataViews = DataViewsRoot as DataViewsProps;
+
+// Expose DataViews components sub components
+DataViews.Search = DataViewsSearch;
```
Just noting that this pattern is going to make these components non tree-shakeable, from my perspective, that's fine since most dataviews are going to use most of the components. 

The alternative would be to export `DataViewsSearch` component on its own... but I personally prefer the current approach used in the PR as these components are highly coupled.

retrofox replied:
Sounds reasonable to me. Going to work on a follow-up to expose the dataviews sub components . Cleaning.

youknowriad reviewed (approved):
This is looking good to me. I'd appreciate a check from @oandregal

retrofox commented:
PR’s merged, André 🙇‍♂️ We can tackle anything else in follow-ups.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/17434026

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @retrofox for including a screenshot in the description! This is really helpful for our translators.

ntsekouras commented on packages/dataviews/src/components/dataviews/index.tsx:
```diff
@@ -112,6 +114,44 @@ export default function DataViews< Item >( {
 	);
 
+	const defaultUI = (
+		<div className="dataviews-wrapper" ref={ containerRef }>
```
Should we keep the wrapper even if `children` are passed? Or alternatively add it in the `DataViewsLayout` component possibly?

The width from `containerRef` is used in `PreviewSizePicker` to dynamically calculate breakpoints and column numbers in `grid` layout. I didn't test, but I wouldn't expect it work properly right now with `free composition`.

retrofox replied:
It's a good question. What is the responsibility of the DataViews in terms of UI in the free composition mode?

In this case, it should be addressed internally, but at a different level, probably in the Layout component.

mcsf commented on packages/dataviews/src/components/dataviews/index.tsx:
```diff
@@ -133,43 +173,7 @@ export default function DataViews< Item >( {
-				<DataViewsLayout />
-				<DataViewsFooter />
-			</div>
+			{ children || defaultUI }
```
It's strange that `defaultUI` is always computed (and recomputed) even if we are passing `children`. That seems wasteful.

youknowriad replied:
I guess it can be a `DefaultUI` component rather than an element.
